gallery/identity/ceed-identity.h

```diff
@@ -21,7 +21,7 @@ CEED_QFUNCTION(Identity)(void *ctx, const CeedInt Q,
                          const CeedScalar *const *in,
                          CeedScalar *const *out) {
   // Ctx holds field size
-  const CeedInt size = ctx ? *(CeedInt *)ctx : 1;
+  const CeedInt size = *(CeedInt *)ctx;
```

Comment:
The ternary really is required here. `ctx` might not be populated with a value. If it is not, then 1 needs to be used.

Reply:
Ah, I see what you did. This works ok on CUDA.... does it not on OCCA?

Reply:
This will force a transfer of data even when `size = 1`. I was trying to skip that for performance.

Reply:
Not sure if it works on CUDA, I can test it out in a few minutes. Maybe runtime CUDA allows for passing `NULL`, but the driver API doesn't seem to like it

Reply:
Yeah, it doesn't seem to work in the regular CUDA backend either. If I run `/gpu/cuda` mode, I get this error:
```
/home/david/git/ceed/backends/cuda/ceed-cuda-vector.c:50 in CeedSyncD2H_Cuda(): CUDA_ERROR_ILLEGAL_ADDRESS
FAIL: t411-qfunction
```

Reply:
Strange, that test passed fine for me on CU's machine.

Reply:
Here's my `nvcc` info:
```
> nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Wed_Oct_23_19:24:38_PDT_2019
Cuda compilation tools, release 10.2, V10.2.89
```

Reply:
That, error isn't even in the Qfunction file..... That's bizzare.

Reply:
Ah, this thread discusses `NULL` with CUDA:

https://what.thedailywtf.com/topic/8661/sometimes-checking-for-null-pointers-is-a-mistake-nvidia-cuda

Looks like it's somewhat architecture dependent :/

Reply:
That's the result of running `make prove search=t411 PROVE_OPTS=v`?

Reply:
I'm running it with a wrapper bash script, how should I run it?

Reply:
```
> prove search=t411 PROVE_OPTS=v
Cannot detect source of 'search=t411'! at /usr/share/perl/5.26/TAP/Parser/IteratorFactory.pm line 261.
	TAP::Parser::IteratorFactory::detect_source(TAP::Parser::IteratorFactory=HASH(0x55a47a221910), TAP::Parser::Source=HASH(0x55a47a0e7d90)) called at /usr/share/perl/5.26/TAP/Parser/IteratorFactory.pm line 211
	TAP::Parser::IteratorFactory::make_iterator(TAP::Parser::IteratorFactory=HASH(0x55a47a221910), TAP::Parser::Source=HASH(0x55a47a0e7d90)) called at /usr/share/perl/5.26/TAP/Parser.pm line 472
	TAP::Parser::_initialize(TAP::Parser=HASH(0x55a47a0e7b08), HASH(0x55a479e75cc8)) called at /usr/share/perl/5.26/TAP/Object.pm line 55
	TAP::Object::new("TAP::Parser", HASH(0x55a479e75cc8)) called at /usr/share/perl/5.26/TAP/Object.pm line 130
	TAP::Object::_construct(TAP::Harness=HASH(0x55a479d9d720), "TAP::Parser", HASH(0x55a479e75cc8)) called at /usr/share/perl/5.26/TAP/Harness.pm line 852
	TAP::Harness::make_parser(TAP::Harness=HASH(0x55a479d9d720), TAP::Parser::Scheduler::Job=HASH(0x55a479e694a0)) called at /usr/share/perl/5.26/TAP/Harness.pm line 651
	TAP::Harness::_aggregate_single(TAP::Harness=HASH(0x55a479d9d720), TAP::Parser::Aggregator=HASH(0x55a4798c50e0), TAP::Parser::Scheduler=HASH(0x55a479e69428)) called at /usr/share/perl/5.26/TAP/Harness.pm line 743
	TAP::Harness::aggregate_tests(TAP::Harness=HASH(0x55a479d9d720), TAP::Parser::Aggregator=HASH(0x55a4798c50e0), "search=t411", "PROVE_OPTS=v") called at /usr/share/perl/5.26/TAP/Harness.pm line 558
	TAP::Harness::__ANON__() called at /usr/share/perl/5.26/TAP/Harness.pm line 571
	TAP::Harness::runtests(TAP::Harness=HASH(0x55a479d9d720), "search=t411", "PROVE_OPTS=v") called at /usr/share/perl/5.26/App/Prove.pm line 546
	App::Prove::_runtests(App::Prove=HASH(0x55a4798b4ca8), HASH(0x55a479d5cde8), "search=t411", "PROVE_OPTS=v") called at /usr/share/perl/5.26/App/Prove.pm line 504
	App::Prove::run(App::Prove=HASH(0x55a4798b4ca8)) called at /usr/bin/prove line 13
```

Reply:
do you have time to get on Google Hangouts to iterate on this faster?

Reply:
It's a minor performance issue, so it's not a big deal to always copy the size to the device. It's really bizarre that changing the source of a Qfunction would give an error in code for a vector.

Reply:
It is unfortunately very common, because kernel calls are asynchronous, we often detect errors not at the end of a kernel but when launching the next one.

Reply:
@YohannDudouit, I think this goes on the list of reasons why having constant and dynamic context would be nice.

Though, for identity restrictions we'll want eventually to put in the code to skip the actual QFunction body and just use the input arrays as output arrays (I now think I might be able to figure that out for CUDA, its in CPU already). It will improve our multigrid performance (once diagonal assembly works on GPU)

Reply:
Added checking for the identity qfunction here:

https://github.com/CEED/libCEED/blob/3ff35f3185f692005ea524a61e482bc93223dfad/backends/occa/operator.hpp#L49

Will try to use this! The CPU ref example made it look easy :crossed_fingers: